### PR TITLE
Transfer learning

### LINKED ISSUE #7: Create fine-tuning script
Script for fine tuning of a model. 
Similar code structure with `train function.
`https://github.com/tyiannak/deep_audio_features/blob/0b2c4b6c534e6dc23f7a5c74471789a88362dadf/core/basic_training.py#L20

### PR DESCRIPTION
**Pull request by DimitrisKatsiros**
Fixes #7 .

### CODE REVIEW

**Comment by tyiannak** [comment]
@lobracost also review this PR whenever ur available

**Comment by tyiannak** [comment]
@DimitrisKatsiros please dont make any changes in the next two days we may push a couple of minor changes in this branch.

**Comment by DimitrisKatsiros** [comment]
@tyiannak ok

**Comment by lobracost** [comment]
I tried to run the code.

At first I a had a problem running it, since the output size of the cnn was not dynamically calculated. I fixed it by introducing a class method for the cnn model, named calc_out_size and commited the changes. Now the model can run for any value of max_sequence_length of a dataset. 

Here I have to note that the pkl folder was missing at this branch.

After that, the training and testing scripts ran as expected.

As I took a look at the code, everything seems good. I can make some suggestions for future improvement:

1. training.py:

- After two bad (increasing loss), or neutral (non changing loss)  epochs the training is finished. This choice might get you stucked at local minimas. A greater tolerance, for example a tolerance of 10 epoch,. would be better.

- I suggest to use a learning rate scheduler (e.g. torch.optim.lr_scheduler.ReduceLROnPlateau) in order to optimize your training process.

2. test.py:
It would be useful if we had the ability to test a set of input data and not just a simple file, so we could get some statistics.

3. 
I recommend to structure the code in a package form with __init__.py files. In this way there is no need to mess with the python path variable.

I will now check the Transfer Learning functionality.

 training.py

**Comment by lobracost** [comment]
The Transfer Learning script runs well too.

I tried to test the model on some data @tyiannak provided me. This data contains 4 classes (music, other, silence, speech).

I first trained a model (model_1) to classify the classes other & silence. 
Then I kept a small part of the classes music & speech, since this is the usual case to use transfer learning.
I trained a new model (model_2) to these classes and the outcome was: 

- Accuracy for first class: 100%

- Accuracy for second class: 65.5%

I then trained model_1 to classify music & speech, using TL.
The outcome on the same test data was:

- Accuracy for first class: 98%

- Accuracy for second class: 78.6%

Seems that transfer learning works as expected and improves the performance when small training dataset is provided.

**Comment by DimitrisKatsiros** [comment]
@lobracost thank you so much for the comments! 
@tyiannak Should I open a new issue / branch for each of the suggestions above ?

**Comment by tyiannak** [comment]
@lobracost thanx for the contribution! I will finish my review in a while and merge if everything is ok.
@DimitrisKatsiros i've created a new issue (9), so we can create a new branch as a next step. (also @lobracost and @DimitrisKatsiros please add in issue 9 any additional subtasks that could be handled in the context of this refactoring task - as a comment in the issue itself)

**Comment by tyiannak** [comment]
ok i did some minor changes in this branch and tested end2end all train,test and transfer scripts.
I am ok, but @lobracost you can do a final check after my changes and do the final merging ok? I am approving.

**Comment by lobracost** [comment]
> ok i did some minor changes in this branch and tested end2end all train,test and transfer scripts.
> I am ok, but @lobracost you can do a final check after my changes and do the final merging ok? I am approving.

Everything works as expected. I'm merging the branch.
